engine/lib/phases/phase_rewrite_control_flow.ml

```diff
@@ -11,31 +14,127 @@ module Make (F : Features.T) =
 
       open Ast.Make (F)
       module U = Ast_utils.Make (F)
+      module M = Ast_builder.Make (F)
       module Visitors = Ast_visitors.Make (F)
 
       module Error = Phase_utils.MakeError (struct
         let ctx = Diagnostics.Context.Phase phase_id
       end)
 
       let has_return =
+        object (_self)
+          inherit [_] Visitors.reduce as super
+          method zero = false
+          method plus = ( || )
+
+          method! visit_expr' in_loop e =
+            match e with
+            | Return _ -> true
+            | (Break _ | Continue _) when in_loop -> true
+            | _ -> super#visit_expr' in_loop e
+        end
```

Comment:
Why is that OK to return true if we face e.g. `loop { loop { return 3 }}`? I mean, shouldn't `has_return` return true only if there is a return in the current loop/function?
If it is OK, can you explain with a comment?

Reply:
Even nested `return`s apply to the top level loop but `break`s and `continue`s don't. I'll add a comment.